<a id = 'title'></a>
# GoodReads Dataset EDA

#### Author: Quinci Birker

### Introduction

### Data Dictionary

| Attributes  | Definition | Completeness |
| ------------- | ------------- | ------------- | 
| bookId  | Book Identifier as in goodreads.com  | 100 |
| title  | Book title | 100 |
| series | Series Name | 45 |
| author | Book's Author | 100 |
| rating | Global goodreads rating | 100 |
| description | Book's description | 97 |
| language | Book's language | 93 |
| isbn | Book's ISBN | 92 |
| genres | Book's genres | 91 |
| characters | Main characters | 26 |
| bookFormat | Type of binding | 97 |
| edition | Type of edition (ex. Anniversary Edition) | 9 |
| pages | Number of pages | 96 |
| publisher | Editorial | 93 |
| publishDate | publication date | 98 |
| firstPublishDate | Publication date of first edition | 59 |
| awards | List of awards | 20 |
| numRatings | Number of total ratings | 100 |
| ratingsByStars | Number of ratings by stars | 97 |
| likedPercent | Derived field, percent of ratings over 2 starts (as in GoodReads) | 99 |
| setting | Story setting | 22 |
| coverImg | URL to cover image | 99 |
| bbeScore | Score in Best Books Ever list | 100 |
| bbeVotes | Number of votes in Best Books Ever list | 100 |
| price | Book's price (extracted from Iberlibro) | 73 |


### Table of Contents
---------------------------------------
- [Import Dataset and Libraries](#import)
---------------------------------------
- [Preliminary Data Exploration](#pre_explore)
---------------------------------------
- [Data Cleaning](#data_clean)
---------------------------------------
- [Descriptive Statistics](#desc_statistics)
---------------------------------------
- [Data Visualization](#data_visualization)
---------------------------------------
- [Correlation Analysis](#correlation_analysis)
---------------------------------------
- [Summary & Insights](#summary)
---------------------------------------


<a id = 'import'></a>
### Import Dataset and Libraries

In [39]:
# import libraries:
import pandas as pd
import numpy as np
import plotly.express as px

In [40]:
# read data from the CSV file:
raw_df = pd.read_csv('data/books_1.Best_Books_Ever.csv')

<a id = 'pre_explore'></a>
### Preliminary Data Exploration

In [41]:
# check number of rows and columns:
raw_df.shape
print(f'There are {raw_df.shape[0]} rows and {raw_df.shape[1]} columns in the data')

There are 52478 rows and 25 columns in the data


In [42]:
# check the data types for every column:
raw_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52478 entries, 0 to 52477
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   bookId            52478 non-null  object 
 1   title             52478 non-null  object 
 2   series            23470 non-null  object 
 3   author            52478 non-null  object 
 4   rating            52478 non-null  float64
 5   description       51140 non-null  object 
 6   language          48672 non-null  object 
 7   isbn              52478 non-null  object 
 8   genres            52478 non-null  object 
 9   characters        52478 non-null  object 
 10  bookFormat        51005 non-null  object 
 11  edition           4955 non-null   object 
 12  pages             50131 non-null  object 
 13  publisher         48782 non-null  object 
 14  publishDate       51598 non-null  object 
 15  firstPublishDate  31152 non-null  object 
 16  awards            52478 non-null  object

The majority of these columns are object. Columns that might need to change from object to integer:
- publish date
- first publish date
- price

In [43]:
# sanity check the first five rows:
raw_df.head()

bookId  \
0                     2767052-the-hunger-games   
1  2.Harry_Potter_and_the_Order_of_the_Phoenix   
2                   2657.To_Kill_a_Mockingbird   
3                     1885.Pride_and_Prejudice   
4                               41865.Twilight   

                                       title                 series  \
0                           The Hunger Games    The Hunger Games #1   
1  Harry Potter and the Order of the Phoenix        Harry Potter #5   
2                      To Kill a Mockingbird  To Kill a Mockingbird   
3                        Pride and Prejudice                    NaN   
4                                   Twilight   The Twilight Saga #1   

                                      author  rating  \
0                            Suzanne Collins    4.33   
1  J.K. Rowling, Mary GrandPré (Illustrator)    4.50   
2                                 Harper Lee    4.28   
3  Jane Austen, Anna Quindlen (Introduction)    4.26   
4                            Stephenie Meyer    3.60   

                                         description language           isbn  \
0  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...  English  9780439023481   
1  There is a door at the end of a silent corrido...  English  9780439358071   
2  The unforgettable novel of a childhood in a sl...  English  9999999999999   
3  Alternate cover edition of ISBN 9780679783268S...  English  9999999999999   
4  About three things I was absolutely positive.\...  English  9780316015844   

                                              genres  \
0  ['Young Adult', 'Fiction', 'Dystopia', 'Fantas...   
1  ['Fantasy', 'Young Adult', 'Fiction', 'Magic',...   
2  ['Classics', 'Fiction', 'Historical Fiction', ...   
3  ['Classics', 'Fiction', 'Romance', 'Historical...   
4  ['Young Adult', 'Fantasy', 'Romance', 'Vampire...   

                                          characters  ... firstPublishDate  \
0  ['Katniss Everdeen', 'Peeta Mellark', 'Cato (H...  ...              NaN   
1  ['Sirius Black', 'Draco Malfoy', 'Ron Weasley'...  ...         06/21/03   
2  ['Scout Finch', 'Atticus Finch', 'Jem Finch', ...  ...         07/11/60   
3  ['Mr. Bennet', 'Mrs. Bennet', 'Jane Bennet', '...  ...         01/28/13   
4  ['Edward Cullen', 'Jacob Black', 'Laurent', 'R...  ...         10/05/05   

                                              awards numRatings  \
0  ['Locus Award Nominee for Best Young Adult Boo...    6376780   
1  ['Bram Stoker Award for Works for Young Reader...    2507623   
2  ['Pulitzer Prize for Fiction (1961)', 'Audie A...    4501075   
3                                                 []    2998241   
4  ['Georgia Peach Book Award (2007)', 'Buxtehude...    4964519   

                                      ratingsByStars likedPercent  \
0  ['3444695', '1921313', '745221', '171994', '93...         96.0   
1  ['1593642', '637516', '222366', '39573', '14526']         98.0   
2  ['2363896', '1333153', '573280', '149952', '80...         95.0   
3  ['1617567', '816659', '373311', '113934', '767...         94.0   
4  ['1751460', '1113682', '1008686', '542017', '5...         78.0   

                                             setting  \
0  ['District 12, Panem', 'Capitol, Panem', 'Pane...   
1  ['Hogwarts School of Witchcraft and Wizardry (...   
2               ['Maycomb, Alabama (United States)']   
3  ['United Kingdom', 'Derbyshire, England (Unite...   
4  ['Forks, Washington (United States)', 'Phoenix...   

                                            coverImg  bbeScore bbeVotes  price  
0  https://i.gr-assets.com/images/S/compressed.ph...   2993816    30516   5.09  
1  https://i.gr-assets.com/images/S/compressed.ph...   2632233    26923   7.38  
2  https://i.gr-assets.com/images/S/compressed.ph...   2269402    23328    NaN  
3  https://i.gr-assets.com/images/S/compressed.ph...   1983116    20452    NaN  
4  https://i.gr-assets.com/images/S/compressed.ph...   1459448    14874    2.1  

[5 rows x 25 columns]

In [44]:
# check last 5 rows:
raw_df.tail()

bookId                title  \
52473            11492014-fractured            Fractured   
52474              11836711-anasazi              Anasazi   
52475               10815662-marked               Marked   
52476          11330278-wayward-son          Wayward Son   
52477  10991547-daughter-of-helaman  Daughter of Helaman   

                     series  \
52473            Fateful #2   
52474     Sense of Truth #2   
52475     Soul Guardians #1   
52476                   NaN   
52477  Stripling Warrior #1   

                                                  author  rating  \
52473                   Cheri Schmidt (Goodreads Author)    4.00   
52474                                      Emma Michaels    4.19   
52475                  Kim Richardson (Goodreads Author)    3.70   
52476  Tom Pollack (Goodreads Author), John Loftus (G...    3.85   
52477                    Misty Moncur (Goodreads Author)    4.02   

                                             description language  \
52473  The Fateful Trilogy continues with Fractured. ...  English   
52474  'Anasazi', sequel to 'The Thirteenth Chime' by...  English   
52475  --READERS FAVORITE AWARDS WINNER 2011--Sixteen...  English   
52476  A POWERFUL TREMOR UNEARTHS AN ANCIENT SECRETBu...  English   
52477  Fighting in Helaman's army is Keturah's deepes...  English   

                isbn                                             genres  \
52473  2940012616562  ['Vampires', 'Paranormal', 'Young Adult', 'Rom...   
52474  9999999999999                         ['Mystery', 'Young Adult']   
52475  9781461017097  ['Fantasy', 'Young Adult', 'Paranormal', 'Ange...   
52476  9781450755634  ['Fiction', 'Mystery', 'Historical Fiction', '...   
52477  9781599554976  ['Lds Fiction', 'Historical Fiction', 'Young A...   

      characters  ... firstPublishDate  \
52473         []  ...              NaN   
52474         []  ...  August 3rd 2011   
52475         []  ...  March 15th 2011   
52476         []  ...   April 5th 2011   
52477         []  ...              NaN   

                                        awards numRatings  \
52473                                       []        871   
52474                                       []         37   
52475  ["Readers' Favorite Book Award (2011)"]       6674   
52476                                       []        238   
52477                                       []        246   

                               ratingsByStars likedPercent setting  \
52473       ['311', '310', '197', '42', '11']         94.0      []   
52474             ['16', '14', '5', '2', '0']         95.0      []   
52475  ['2109', '1868', '1660', '647', '390']         84.0      []   
52476           ['77', '78', '59', '19', '5']         90.0      []   
52477          ['106', '73', '42', '17', '8']         90.0      []   

                                                coverImg  bbeScore bbeVotes  \
52473  https://i.gr-assets.com/images/S/compressed.ph...         0        1   
52474  https://i.gr-assets.com/images/S/compressed.ph...         0        1   
52475  https://i.gr-assets.com/images/S/compressed.ph...         0        1   
52476  https://i.gr-assets.com/images/S/compressed.ph...         0        1   
52477  https://i.gr-assets.com/images/S/compressed.ph...         0        1   

       price  
52473    NaN  
52474    NaN  
52475   7.37  
52476   2.86  
52477   5.20  

[5 rows x 25 columns]

In [45]:
raw_df.sample(5)

bookId  \
27564                 149704.The_Valley_of_the_Moon   
52444                     96337.The_Conservationist   
22092        842309.Idle_Thoughts_of_an_Idle_Fellow   
33121  644478.Letters_from_a_Father_to_his_Daughter   
23961                     13030276-so-shall-we-pass   

                                       title            series  \
27564                 The Valley of the Moon               NaN   
52444                    The Conservationist               NaN   
22092        Idle Thoughts of an Idle Fellow  Idle Thoughts #1   
33121  Letters from a Father to his Daughter               NaN   
23961                       So Shall We Pass               NaN   

                                   author  rating  \
27564                         Jack London    3.99   
52444                     Nadine Gordimer    3.36   
22092                    Jerome K. Jerome    3.87   
33121                    Jawaharlal Nehru    3.93   
23961  Michael Barrera (Goodreads Author)    4.92   

                                             description language  \
27564  A road novel 50 years before Kerouac, The Vall...  English   
52444  Mehring is rich. He has all the privileges and...  English   
22092  Jerome Klapka Jerome (1859 - 1927), English au...  English   
33121  A priceless collection of letters from one leg...  English   
23961  So Shall We Pass is a first-person present ten...  English   

                isbn                                             genres  \
27564  9781406946253  ['Classics', 'Fiction', 'Literature', 'Adventu...   
52444  9780140047165  ['Fiction', 'Africa', 'South Africa', 'Nobel P...   
22092  9781595690241  ['Humor', 'Nonfiction', 'Classics', 'Essays', ...   
33121  9780670058167  ['Nonfiction', 'History', 'India', 'Indian Lit...   
23961  9781462042661                                                 []   

                               characters  ... firstPublishDate  \
27564  ['Billy Roberts', 'Saxon Roberts']  ...         10/30/13   
52444                                  []  ...             1974   
22092                                  []  ...         10/30/86   
33121                                  []  ...         11/06/29   
23961                                  []  ...         08/01/11   

                                                  awards numRatings  \
27564                                                 []       1395   
52444  ['Booker Prize (1974)', 'The Best of the Booke...       2509   
22092                                                 []       2393   
33121                                                 []       2052   
23961                                                 []         26   

                            ratingsByStars likedPercent  \
27564    ['503', '504', '285', '83', '20']         93.0   
52444  ['389', '772', '825', '394', '129']         79.0   
22092   ['729', '852', '627', '140', '45']         92.0   
33121    ['673', '758', '476', '94', '51']         93.0   
23961           ['24', '2', '0', '0', '0']        100.0   

                              setting  \
27564  ['California (United States)']   
52444                ['South Africa']   
22092                              []   
33121                              []   
23961                              []   

                                                coverImg  bbeScore bbeVotes  \
27564  https://i.gr-assets.com/images/S/compressed.ph...        96        1   
52444  https://i.gr-assets.com/images/S/compressed.ph...         1        1   
22092  https://i.gr-assets.com/images/S/compressed.ph...        99        1   
33121  https://i.gr-assets.com/images/S/compressed.ph...        92        1   
23961  https://i.gr-assets.com/images/S/compressed.ph...        98        1   

       price  
27564  36.65  
52444   1.12  
22092  13.00  
33121   5.64  
23961  10.00  

[5 rows x 25 columns]

Notes from looking at the begining, end, and a random sample of the dataset:

1. Columns to be deleted that are not useful for this project:
    - `bookId` does not seem to be useful for my project. The index will be used instead of this.  
    - `isbn` is a numeric book identifier. I will also delete this since I am using the index to refer to each unique book in the dataset.
    - `coverImg` is a url to the books cover image. For this project, I will not be using this in my modeling.  
2. There are quite a few columns that are missing data. Further analysis will be performed.

3. Columns that stand out for further investigation:
    - `firstPublishDate` ~ missing values and different formats (i.e. 07/29/96, 1989, April 5th 2011)
    - `price` ~ the values don't seem to be accurate. The entire column might need to be deleted

In [46]:
# Check that total index matches total number of rows:
raw_df.index.nunique() == raw_df.shape[0]

True

The total index count is equal to the total number of rows in the dataset. 

In [47]:
# Count the number of missing values for each column:
raw_df.isna().sum()

bookId                  0
title                   0
series              29008
author                  0
rating                  0
description          1338
language             3806
isbn                    0
genres                  0
characters              0
bookFormat           1473
edition             47523
pages                2347
publisher            3696
publishDate           880
firstPublishDate    21326
awards                  0
numRatings              0
ratingsByStars          0
likedPercent          622
setting                 0
coverImg              605
bbeScore                0
bbeVotes                0
price               14365
dtype: int64

In [48]:
# Percentage of missing values for each column:
print(raw_df.isna().sum(axis=0)/raw_df.shape[0])

bookId              0.000000
title               0.000000
series              0.552765
author              0.000000
rating              0.000000
description         0.025496
language            0.072526
isbn                0.000000
genres              0.000000
characters          0.000000
bookFormat          0.028069
edition             0.905579
pages               0.044724
publisher           0.070430
publishDate         0.016769
firstPublishDate    0.406380
awards              0.000000
numRatings          0.000000
ratingsByStars      0.000000
likedPercent        0.011853
setting             0.000000
coverImg            0.011529
bbeScore            0.000000
bbeVotes            0.000000
price               0.273734
dtype: float64


There are 12 columns that have missing values. The column name and rounded percent of missing values in descending order:
- edition: 91%
- series: 55%
- first publish date: 41%
- price 27%
- language: 7%
- publisher: 7%
- pages: 4%
- description: 3%
- book format: 3%
- publish date: 2%
- liked percent: 1%
- cover image: 1%

The first step in the data cleaning process is to delete all the columns that I will not be using for modeling:
- bookid: using dataset index instead
- isbn: using dataset index instead
- cover image: won't be used for this modeling
- edition: over 91% of the values are missing
- first publish date: over 41% missing and my assumption is that publish date will be more relevant than first publish date
- price: the prices do not seem accurate enough to use. Secondly, depending on the date of purchase and retailer, the price of books can very greatly. 
- awards: there is limited values in this dataset
- characters: there is limited values in this dataset
- setting: there is limited values in this dataset

# Come back to this!!

In [49]:
# Check for empty lists in the 'genres' column
empty_lists_mask = raw_df['genres'].apply(lambda x: len(x) == 0)

# Identify rows with empty lists
rows_with_empty_lists = raw_df[empty_lists_mask]

# Display the result
print(rows_with_empty_lists)

Empty DataFrame
Columns: [bookId, title, series, author, rating, description, language, isbn, genres, characters, bookFormat, edition, pages, publisher, publishDate, firstPublishDate, awards, numRatings, ratingsByStars, likedPercent, setting, coverImg, bbeScore, bbeVotes, price]
Index: []

[0 rows x 25 columns]


In [50]:
# Check for duplicated columns:
raw_df.T.duplicated().sum()

0

In [51]:
# Checking for duplicate rows: 
raw_df[raw_df.duplicated()]

bookId  \
37431                           8794263-promises-to-keep   
37432                                      1909590.Click   
37433                 23394408-die-unendlichkeit-schl-ft   
37434                                 7544945-death-note   
37435                                  25886017-m-scaras   
37436                      2669775-el-siglo-de-las-luces   
37437                        24633605-always-and-forever   
37438                           237086.Dafne_desvanecida   
37439                                           17786377   
37440    7452583-jag-vill-inte-d-jag-vill-bara-inte-leva   
37441                                           18480662   
37442                        1258333.Cantarella_Volume_1   
37443                                  17999135-exposure   
37444                           220362.Manon_des_sources   
37445                                           23438427   
37446                                  650584.Shadowdale   
37447                           24644867-traiesc-din-nou   
37448                          22587955-count-with-kenny   
37449             8438589-the-creatures-that-time-forgot   
37450                          17337543-the-game-changer   
37451                       22372599-the-seahorse-legacy   
37452                    9807272-big-nate-boredom-buster   
37454              23590446-how-to-fight-a-dragon-s-fury   
37456                             16058645-the-bookstore   
37457          15733451-chairul-tanjung-si-anak-singkong   
37458                                           25454580   
37459                              3486235-night-letters   
37460                         11737313-three-times-lucky   
37461                              13614864-know-not-why   
37462                              21900147-gone-too-far   
37463                      15741217-writing-scary-scenes   
37464  6665889-the-science-fiction-hall-of-fame-volum...   
37465                                2081377.Ghosthunter   
37466                                  3973924-bal-masqu   
37467                        18849994-beautifully-broken   
37468  8538135-catfancy-s-naughty-no-more-change-unwa...   
37469                 18813012-the-rocker-that-holds-her   
37470                                           17406982   
37471                                 21254063-sorceress   
37472                                25590067-conviction   
37473                           383514.Her_Royal_Spyness   
37474              2074104.The_Legend_of_Colton_H_Bryant   
37475                      20262067-le-festin-du-serpent   
37478                                    23519340-silber   
37479                              29742514-mays-landing   
37480                                26845691-beatlebone   
37481                22916606-gli-enigmi-dei-vedovi-neri   
37482                                         13571757-1   
37483    1481056.Gifts_and_Ministries_of_the_Holy_Spirit   
37484                        27074409-deadly-class-vol-1   

                                                   title  \
37431                                   Promises to Keep   
37432                                              Click   
37433                          Die Unendlichkeit schläft   
37434                  Death Note: Black Edition, Vol. 2   
37435                                           Máscaras   
37436                              El siglo de las luces   
37437                                 Always and Forever   
37438                                  Dafne desvanecida   
37439                                       فضولية العلم   
37440          Jag vill inte dö, jag vill bara inte leva   
37441                     لا سكاكين في مطابخ هذه المدينة   
37442                               Cantarella, Volume 1   
37443                                           Exposure   
37444                                  Manon des sources   
37445                                       পিতা ও পুত্র   
37446                        

These duplicated rows have all the same bbeScore and bbeVotes, yet all the book titles are unique, meaning that these rows will not be dropped.

<a id = 'data_clean'></a>
### Data Cleaning

Step 1: Dropping the columns that will not be used in the analysis/modeling.

In [52]:
# Drop specified column and check that changes have been made to raw_df:

raw_df = raw_df.drop(['isbn', 'coverImg', 'edition', 'firstPublishDate', 'price', 'awards', 'setting', 'characters'], axis=1)
raw_df.head()

bookId  \
0                     2767052-the-hunger-games   
1  2.Harry_Potter_and_the_Order_of_the_Phoenix   
2                   2657.To_Kill_a_Mockingbird   
3                     1885.Pride_and_Prejudice   
4                               41865.Twilight   

                                       title                 series  \
0                           The Hunger Games    The Hunger Games #1   
1  Harry Potter and the Order of the Phoenix        Harry Potter #5   
2                      To Kill a Mockingbird  To Kill a Mockingbird   
3                        Pride and Prejudice                    NaN   
4                                   Twilight   The Twilight Saga #1   

                                      author  rating  \
0                            Suzanne Collins    4.33   
1  J.K. Rowling, Mary GrandPré (Illustrator)    4.50   
2                                 Harper Lee    4.28   
3  Jane Austen, Anna Quindlen (Introduction)    4.26   
4                            Stephenie Meyer    3.60   

                                         description language  \
0  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...  English   
1  There is a door at the end of a silent corrido...  English   
2  The unforgettable novel of a childhood in a sl...  English   
3  Alternate cover edition of ISBN 9780679783268S...  English   
4  About three things I was absolutely positive.\...  English   

                                              genres bookFormat pages  \
0  ['Young Adult', 'Fiction', 'Dystopia', 'Fantas...  Hardcover   374   
1  ['Fantasy', 'Young Adult', 'Fiction', 'Magic',...  Paperback   870   
2  ['Classics', 'Fiction', 'Historical Fiction', ...  Paperback   324   
3  ['Classics', 'Fiction', 'Romance', 'Historical...  Paperback   279   
4  ['Young Adult', 'Fantasy', 'Romance', 'Vampire...  Paperback   501   

                          publisher publishDate  numRatings  \
0                  Scholastic Press    09/14/08     6376780   
1                   Scholastic Inc.    09/28/04     2507623   
2  Harper Perennial Modern Classics    05/23/06     4501075   
3                    Modern Library    10/10/00     2998241   
4         Little, Brown and Company    09/06/06     4964519   

                                      ratingsByStars  likedPercent  bbeScore  \
0  ['3444695', '1921313', '745221', '171994', '93...          96.0   2993816   
1  ['1593642', '637516', '222366', '39573', '14526']          98.0   2632233   
2  ['2363896', '1333153', '573280', '149952', '80...          95.0   2269402   
3  ['1617567', '816659', '373311', '113934', '767...          94.0   1983116   
4  ['1751460', '1113682', '1008686', '542017', '5...          78.0   1459448   

   bbeVotes  
0     30516  
1     26923  
2     23328  
3     20452  
4     14874

Step 2: Reformatting the publish date column.

For the publish date, the first 30,000 books in the dataset are formatted in mm/dd/yyyy while the last 22,478 books are formated in Month Day Year. I will reformate all the dates to be in mm/dd/yyyy.

In [53]:
# Change publish date to datetime format:
raw_df['publishDate'] = pd.to_datetime(raw_df['publishDate'], errors='coerce') # errors='coerce' used if value cannot be converted to datetime format
# Reformat publish date to mm/dd/yyyy format:
raw_df['publishDate'] = raw_df['publishDate'].dt.strftime('%m/%d/%Y')

/var/folders/b1/ck799gjn11b_0zbx9y3pynvr0000gn/T/ipykernel_36851/2598813278.py:2: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  raw_df['publishDate'] = pd.to_datetime(raw_df['publishDate'], errors='coerce') # errors='coerce' used if value cannot be converted to datetime format


Sanity check that this change was made to the dataset:

In [54]:
raw_df.sample(5)

bookId  \
7681                 31818.The_Philosophy_of_Andy_Warhol   
160                            7190.The_Three_Musketeers   
38096                       18148148-writing-about-magic   
36633  21253714-the-alchemy-press-book-of-ancient-won...   
8519                      26198109-louder-than-a-whisper   

                                                   title  \
7681   The Philosophy of Andy Warhol (From A to B and...   
160                                 The Three Musketeers   
38096                                Writing about Magic   
36633          The Alchemy Press Book of Ancient Wonders   
8519          Louder Than a Whisper: Clearer Than a Bell   

                                series  \
7681                               NaN   
160         The d'Artagnan Romances #1   
38096                Writer's Craft #3   
36633  Tales of the Apt #Story - Bones   
8519                               NaN   

                                                  author  rating  \
7681                                         Andy Warhol    3.77   
160                                      Alexandre Dumas    4.07   
38096                                         Rayne Hall    3.87   
36633  Jan Edwards (Goodreads Author) (Editor), Jenny...    3.86   
8519        Renée Paule (Goodreads Author) (Illustrator)    4.71   

                                             description language  \
7681   A loosely formed autobiography by Andy Warhol,...  English   
160    Alexandre Dumas’s most famous tale— and possib...  English   
38096  Do you write fantasy fiction? This book is a r...      NaN   
36633  When we think of a wonder, our minds go most o...  English   
8519   This book challenges the status quo of humanit...  English   

                                                  genres      bookFormat  \
7681   ['Art', 'Nonfiction', 'Philosophy', 'Biography...       Paperback   
160    ['Classics', 'Fiction', 'Historical Fiction', ...       Paperback   
38096             ['Writing', 'Nonfiction', 'Reference']  Kindle Edition   
36633                                                 []  Kindle Edition   
8519   ['Nonfiction', 'Self Help', 'Inspirational', '...       Paperback   

      pages          publisher publishDate  numRatings  \
7681    272      Mariner Books  04/06/1977       37274   
160     625     Modern Library  02/13/2001      273714   
38096   133                NaN         NaN         150   
36633   252  The Alchemy Press  02/03/2013          21   
8519    174     RPG Publishing  09/21/2015          35   

                                       ratingsByStars  likedPercent  bbeScore  \
7681       ['13336', '10134', '8141', '3104', '2559']          85.0       345   
160    ['103957', '101696', '54638', '10239', '3184']          95.0     95549   
38096                    ['51', '51', '33', '8', '7']          90.0        86   
36633                       ['8', '6', '4', '2', '1']          86.0        88   
8519                       ['28', '4', '3', '0', '0']         100.0       294   

       bbeVotes  
7681          4  
160        1198  
38096         1  
36633         1  
8519          3

# COME BACK TO THIS:

Step 3: Looking at series which is missing 55% of it's values.

In [55]:
raw_df.loc[raw_df['series'].isna()]

bookId  \
3                1885.Pride_and_Prejudice   
5                    19063.The_Book_Thief   
6                      170448.Animal_Farm   
9                18405.Gone_with_the_Wind   
10        11870085-the-fault-in-our-stars   
...                                   ...   
52463  25876358-the-natural-way-of-things   
52464                  36374396-algedonic   
52469               270435.Heal_Your_Body   
52470          11115191-attracted-to-fire   
52476                11330278-wayward-son   

                                                   title series  \
3                                    Pride and Prejudice    NaN   
5                                         The Book Thief    NaN   
6                                            Animal Farm    NaN   
9                                     Gone with the Wind    NaN   
10                                The Fault in Our Stars    NaN   
...                                                  ...    ...   
52463                          The Natural Way of Things    NaN   
52464                                          Algedonic    NaN   
52469  Heal Your Body: The Mental Causes for Physical...    NaN   
52470                                  Attracted to Fire    NaN   
52476                                        Wayward Son    NaN   

                                                  author  rating  \
3              Jane Austen, Anna Quindlen (Introduction)    4.26   
5                        Markus Zusak (Goodreads Author)    4.37   
6      George Orwell, Russell Baker (Preface), C.M. W...    3.95   
9                                      Margaret Mitchell    4.30   
10                         John Green (Goodreads Author)    4.21   
...                                                  ...     ...   
52463                                     Charlotte Wood    3.53   
52464                        R.H. Sin (Goodreads Author)    3.71   
52469                                      Louise L. Hay    4.36   
52470                     DiAnn Mills (Goodreads Author)    4.14   
52476  Tom Pollack (Goodreads Author), John Loftus (G...    3.85   

                                             description language  \
3      Alternate cover edition of ISBN 9780679783268S...  English   
5      Librarian's note: An alternate cover edition c...  English   
6      Librarian's note: There is an Alternate Cover ...  English   
9      Scarlett O'Hara, the beautiful, spoiled daught...  English   
10     Despite the tumor-shrinking medical miracle th...  English   
...                                                  ...      ...   
52463  Two women awaken from a drugged sleep to find ...  English   
52464  Bestselling poet r.h. Sin, author of the Whisk...      NaN   
52469  Heal Your Body is a fresh and easy step-by-ste...  English   
52470  Special Agent Meghan Connors' dream of one day...  English   
52476  A POWERFUL TREMOR UNEARTHS AN ANCIENT SECRETBu...  English   

                                                  genres  \
3      ['Classics', 'Fiction', 'Romance', 'Historical...   
5      ['Historical Fiction', 'Fiction', 'Young Adult...   
6      ['Classics', 'Fiction', 'Dystopia', 'Fantasy',...   
9      ['Classics', 'Historical Fiction', 'Fiction', ...   
10     ['Young Adult', 'Romance', 'Fiction', 'Contemp...   
...                                                  ...   
52463  ['Fiction', 'Dystopia', 'Australia', 'Feminism...   
52464    ['Poetry', 'Nonfiction', 'Romance', 'Feminism']   
52469  ['Self Help', 'Health', 'Nonfiction', 'Spiritu...   
52470  ['Christian Fiction', 'Christian', 'Suspense',...   
52476  ['Fiction', 'Mystery', 'Historical Fiction', '...   

                  bookFormat pages                  publisher publishDate  \
3                  Paperback   279             Modern Library  10/10/2000   
5                  Hardcover   552            Alfred A. Knopf  03/14/2006   
6      Mass Market Paperback   141            Signet Classics  04/28/1996   
9      Mass Ma

In order to use the series in the modeling, I will transform the data into either has a series (value =1) or does not have a series (value=0)

In [58]:
# Fills all NaN values with 0 and then any values that are not 0 are now 1
raw_df['series'] = raw_df['series'].fillna(0).apply(lambda x: 1 if x != 0 else 0)

# Confirm tha these changes were made:
raw_df.head()

bookId  \
0                     2767052-the-hunger-games   
1  2.Harry_Potter_and_the_Order_of_the_Phoenix   
2                   2657.To_Kill_a_Mockingbird   
3                     1885.Pride_and_Prejudice   
4                               41865.Twilight   

                                       title  series  \
0                           The Hunger Games       1   
1  Harry Potter and the Order of the Phoenix       1   
2                      To Kill a Mockingbird       1   
3                        Pride and Prejudice       0   
4                                   Twilight       1   

                                      author  rating  \
0                            Suzanne Collins    4.33   
1  J.K. Rowling, Mary GrandPré (Illustrator)    4.50   
2                                 Harper Lee    4.28   
3  Jane Austen, Anna Quindlen (Introduction)    4.26   
4                            Stephenie Meyer    3.60   

                                         description language  \
0  WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...  English   
1  There is a door at the end of a silent corrido...  English   
2  The unforgettable novel of a childhood in a sl...  English   
3  Alternate cover edition of ISBN 9780679783268S...  English   
4  About three things I was absolutely positive.\...  English   

                                              genres bookFormat pages  \
0  ['Young Adult', 'Fiction', 'Dystopia', 'Fantas...  Hardcover   374   
1  ['Fantasy', 'Young Adult', 'Fiction', 'Magic',...  Paperback   870   
2  ['Classics', 'Fiction', 'Historical Fiction', ...  Paperback   324   
3  ['Classics', 'Fiction', 'Romance', 'Historical...  Paperback   279   
4  ['Young Adult', 'Fantasy', 'Romance', 'Vampire...  Paperback   501   

                          publisher publishDate  numRatings  \
0                  Scholastic Press  09/14/2008     6376780   
1                   Scholastic Inc.  09/28/2004     2507623   
2  Harper Perennial Modern Classics  05/23/2006     4501075   
3                    Modern Library  10/10/2000     2998241   
4         Little, Brown and Company  09/06/2006     4964519   

                                      ratingsByStars  likedPercent  bbeScore  \
0  ['3444695', '1921313', '745221', '171994', '93...          96.0   2993816   
1  ['1593642', '637516', '222366', '39573', '14526']          98.0   2632233   
2  ['2363896', '1333153', '573280', '149952', '80...          95.0   2269402   
3  ['1617567', '816659', '373311', '113934', '767...          94.0   1983116   
4  ['1751460', '1113682', '1008686', '542017', '5...          78.0   1459448   

   bbeVotes  
0     30516  
1     26923  
2     23328  
3     20452  
4     14874

Step 4: Language column has 3,806 (7 percent) missing values. Further investigation will happen here:

In [ ]:
# Look at the rows where the values for the language column are missing:
raw_df.loc[raw_df['language'].isna()]

Step 5: The pages column is missing around 4% of it's data. Further investigation...

In [62]:
# Look at the rows with pages column missing:
raw_df.loc[raw_df['pages'].isna()]

bookId  \
669                                 30346601-brainwalker   
1062                               44575575-pleasant-day   
1428                             318525.Red_Storm_Rising   
1483                              26030383-beg-for-mercy   
1543                       19181419-a-bird-without-wings   
...                                                  ...   
52295         53946399-launching-your-real-estate-career   
52352                               971322.The_Wanderers   
52359                           18130046-ender-in-flight   
52407  2823038-forever-young-forever-free-forever-you...   
52411                                21691391-pixie-dust   

                                                   title  series  \
669                                          Brainwalker       0   
1062                                        Pleasant Day       0   
1428                                    Red Storm Rising       0   
1483                                       Beg For Mercy       1   
1543                                A Bird Without Wings       0   
...                                                  ...     ...   
52295  Launching Your Real Estate Career: The Definit...       0   
52352                                      The Wanderers       0   
52359                                    Ender in Flight       1   
52407  Forever Young, Forever Free Forever Young, For...       0   
52411                                         Pixie Dust       1   

                                                  author  rating  \
669    Robyn Mundell (Goodreads Author), Stephan Laca...    4.35   
1062                   Vera Jane Cook (Goodreads Author)    4.33   
1428                                          Tom Clancy    4.17   
1483                      Lucian Bane (Goodreads Author)    4.29   
1543                   Roberta Pearce (Goodreads Author)    4.17   
...                                                  ...     ...   
52295                                       Eric Kistner    5.00   
52352                                      Richard Price    3.90   
52359                                   Orson Scott Card    4.35   
52407                                       Hettie Jones    3.67   
52411                       Laura Lee (Goodreads Author)    3.79   

                                             description language  \
669    One teen’s incredible journey may just blow hi...  English   
1062   In the town of Hollow Creek, South Carolina tw...  English   
1428   "Allah!"With that shrill cry, three Muslim ter...  English   
1483   The fight is on in this installment, Mercy is ...  English   
1543   After an impoverished and indigent childhood, ...  English   
...                                                  ...      ...   
52295                                                NaN      NaN   
52352  The Wanderers, a teen-age gang in the Bronx of...  English   
52359  "Ender in Flight" is a story by Orson Scott Ca...  English   
52407                                                NaN      NaN   
52411  *A lonesome fairy with no clue how to wield he...      NaN   

                                                  genres  \
669    ['Young Adult', 'Adventure', 'Fantasy', 'Middl...   
1062   ['Contemporary', 'Drama', 'Book Club', 'Fictio...   
1428   ['Fiction', 'Thriller', 'Military Fiction', 'W...   
1483                                                  []   
1543   ['Romance', 'Contemporary Romance', 'Contempor...   
...                                                  ...   
52295                                                 []   
52352  ['Fiction', 'Crime', 'New York', 'Novels', 'Li...   
52359  ['Science Fiction', 'Fiction', 'Young Adult', ...   
52407                                                 []   
52411  ['Fantasy', 'Paranormal', 'Paranormal Romance'...   

                  bookFormat pages            publisher publishDate  \
669                    ebook   NaN  Dualmind Publishing  10/01/2016   
106

For book format and description, I will replace the empty values with 'unknown'. These columns have a low percent of missing values, both being under 3 percent, therefore, this should not have a huge impact on the data. 

In [64]:
raw_df['description'].fillna('unknown', inplace=True)
raw_df['bookFormat'].fillna('unknown', inplace=True)

Last step before moving onto descriptive statistics, check that the empty values have all been dealt with:

In [65]:
# Percentage of missing values for each column:
print(raw_df.isna().sum(axis=0)/raw_df.shape[0])

bookId            0.000000
title             0.000000
series            0.000000
author            0.000000
rating            0.000000
description       0.000000
language          0.072526
genres            0.000000
bookFormat        0.000000
pages             0.044724
publisher         0.070430
publishDate       0.034014
numRatings        0.000000
ratingsByStars    0.000000
likedPercent      0.011853
bbeScore          0.000000
bbeVotes          0.000000
dtype: float64


<a id = 'desc_statistics'></a>
### Descriptive Statistics

Compute the basic statistics of all the numeric columns:

In [28]:
df = raw_df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52478 entries, 0 to 52477
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   bookId          52478 non-null  object 
 1   title           52478 non-null  object 
 2   series          23470 non-null  object 
 3   author          52478 non-null  object 
 4   rating          52478 non-null  float64
 5   description     51140 non-null  object 
 6   language        48672 non-null  object 
 7   genres          52478 non-null  object 
 8   bookFormat      51005 non-null  object 
 9   pages           50131 non-null  object 
 10  publisher       48782 non-null  object 
 11  publishDate     50693 non-null  object 
 12  numRatings      52478 non-null  int64  
 13  ratingsByStars  52478 non-null  object 
 14  likedPercent    51856 non-null  float64
 15  bbeScore        52478 non-null  int64  
 16  bbeVotes        52478 non-null  int64  
dtypes: float64(2), int64(3), object

In [27]:
df.describe()

rating    numRatings  likedPercent      bbeScore      bbeVotes
count  52478.000000  5.247800e+04  51856.000000  5.247800e+04  52478.000000
mean       4.021878  1.787865e+04     92.231545  1.984023e+03     22.529003
std        0.367146  1.039448e+05      5.990689  3.515314e+04    369.158541
min        0.000000  0.000000e+00      0.000000  0.000000e+00     -4.000000
25%        3.820000  3.410000e+02     90.000000  8.400000e+01      1.000000
50%        4.030000  2.307000e+03     94.000000  9.700000e+01      1.000000
75%        4.230000  9.380500e+03     96.000000  1.870000e+02      2.000000
max        5.000000  7.048471e+06    100.000000  2.993816e+06  30516.000000

<a id = 'data_visualization'></a>
### Data Visualization

<a id = 'correlation_analysis'></a>
### Correlation Analysis

<a id = 'summary'></a>
### Summary & Insights

[Back to the top](#title)